File: src/lib/usage.ts
/**
 * Token and cost accounting.
 *
 * TEACHING NOTE: `usage` is the single most under-used field in the API.
 * Four numbers come back on every response and they mean different things:
 *
 *   input_tokens                 uncached input, billed at full rate
 *   cache_creation_input_tokens  written to cache, billed at ~1.25x
 *   cache_read_input_tokens      served from cache, billed at ~0.1x
 *   output_tokens                generated, billed at the output rate
 *
 * The number people expect to be "total input" is the SUM of the first three.
 * If you log only `input_tokens` on a cached workload, your dashboard will
 * show costs collapsing to near zero and you will not notice a broken cache.
 *
 * The single diagnostic that matters: if `cache_read_input_tokens` is 0 across
 * repeated requests that should share a prefix, something is invalidating it.
 */
import { PRICING } from "../config.js";

export interface UsageLike {
  input_tokens: number;
  output_tokens: number;
  cache_creation_input_tokens?: number | null;
  cache_read_input_tokens?: number | null;
}

export interface UsageReport {
  input_tokens: number;
  output_tokens: number;
  cache_creation_input_tokens: number;
  cache_read_input_tokens: number;
  total_input_tokens: number;
  /** true once the cache is warm — the assertion your cache lab checks. */
  cache_hit: boolean;
  estimated_cost_usd: number;
  /** What this call would have cost with no caching at all. */
  uncached_cost_usd: number;
  savings_usd: number;
}

export function summarizeUsage(usage: UsageLike): UsageReport {
  const cacheWrite = usage.cache_creation_input_tokens ?? 0;
  const cacheRead = usage.cache_read_input_tokens ?? 0;
  const fresh = usage.input_tokens;
  const totalInput = fresh + cacheWrite + cacheRead;

  const inRate = PRICING.inputPerMTok / 1_000_000;
  const outRate = PRICING.outputPerMTok / 1_000_000;

  const cost =
    fresh * inRate +
    cacheWrite * inRate * PRICING.cacheWriteMultiplier +
    cacheRead * inRate * PRICING.cacheReadMultiplier +
    usage.output_tokens * outRate;

  const uncached = totalInput * inRate + usage.output_tokens * outRate;

  return {
    input_tokens: fresh,
    output_tokens: usage.output_tokens,
    cache_creation_input_tokens: cacheWrite,
    cache_read_input_tokens: cacheRead,
    total_input_tokens: totalInput,
    cache_hit: cacheRead > 0,
    estimated_cost_usd: round6(cost),
    uncached_cost_usd: round6(uncached),
    savings_usd: round6(uncached - cost),
  };
}

/** Adds usage reports across the many turns of an agentic loop. */
export function sumUsage(reports: UsageReport[]): UsageReport {
  const zero: UsageReport = {
    input_tokens: 0,
    output_tokens: 0,
    cache_creation_input_tokens: 0,
    cache_read_input_tokens: 0,
    total_input_tokens: 0,
    cache_hit: false,
    estimated_cost_usd: 0,
    uncached_cost_usd: 0,
    savings_usd: 0,
  };
  const total = reports.reduce((acc, r) => ({
    input_tokens: acc.input_tokens + r.input_tokens,
    output_tokens: acc.output_tokens + r.output_tokens,
    cache_creation_input_tokens:
      acc.cache_creation_input_tokens + r.cache_creation_input_tokens,
    cache_read_input_tokens: acc.cache_read_input_tokens + r.cache_read_input_tokens,
    total_input_tokens: acc.total_input_tokens + r.total_input_tokens,
    cache_hit: acc.cache_hit || r.cache_hit,
    estimated_cost_usd: acc.estimated_cost_usd + r.estimated_cost_usd,
    uncached_cost_usd: acc.uncached_cost_usd + r.uncached_cost_usd,
    savings_usd: acc.savings_usd + r.savings_usd,
  }), zero);

  total.estimated_cost_usd = round6(total.estimated_cost_usd);
  total.uncached_cost_usd = round6(total.uncached_cost_usd);
  total.savings_usd = round6(total.savings_usd);
  return total;
}

function round6(n: number): number {
  return Math.round(n * 1e6) / 1e6;
}

File: python/triage/usage.py
"""Token and cost accounting. Mirrors src/lib/usage.ts.

Total input is the SUM of three fields, not just `input_tokens`. Log only the
first and a cached workload looks almost free right up until the cache breaks.
"""

from dataclasses import dataclass, asdict

from .config import spec_for


@dataclass
class UsageReport:
    model: str
    batch: bool
    input_tokens: int
    output_tokens: int
    cache_creation_input_tokens: int
    cache_read_input_tokens: int
    total_input_tokens: int
    cache_hit: bool
    estimated_cost_usd: float
    uncached_cost_usd: float
    savings_usd: float

    def dict(self) -> dict:
        return asdict(self)


def summarize_usage(usage, model: str, batch: bool = False) -> UsageReport:
    cache_write = getattr(usage, "cache_creation_input_tokens", 0) or 0
    cache_read = getattr(usage, "cache_read_input_tokens", 0) or 0
    fresh = usage.input_tokens
    total_input = fresh + cache_write + cache_read

    pricing = spec_for(model)
    discount = pricing.batch_multiplier if batch else 1.0
    in_rate = (pricing.input_per_mtok / 1_000_000) * discount
    out_rate = (pricing.output_per_mtok / 1_000_000) * discount

    cost = (
        fresh * in_rate
        + cache_write * in_rate * pricing.cache_write_multiplier
        + cache_read * in_rate * pricing.cache_read_multiplier
        + usage.output_tokens * out_rate
    )
    uncached = total_input * in_rate + usage.output_tokens * out_rate

    r6 = lambda n: round(n, 6)
    return UsageReport(
        model=model,
        batch=batch,
        input_tokens=fresh,
        output_tokens=usage.output_tokens,
        cache_creation_input_tokens=cache_write,
        cache_read_input_tokens=cache_read,
        total_input_tokens=total_input,
        cache_hit=cache_read > 0,
        estimated_cost_usd=r6(cost),
        uncached_cost_usd=r6(uncached),
        savings_usd=r6(uncached - cost),
    )
